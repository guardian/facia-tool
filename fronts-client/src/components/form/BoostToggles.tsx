import InputRadio from 'components/inputs/InputRadio';
import React from 'react';
import { Field } from 'redux-form';
import type { BoostLevels } from 'types/Collection';

/**
 * This function generates a set of radio button toggles for selecting a card's "boost level"
 * in a flexible collection. Boost levels help establish a visual hierarchy within a collection.
 *
 * ## Boost Levels:
 * - **default**: No boost
 * - **boost**: Slight emphasis
 * - **megaBoost**: Stronger emphasis
 * - **gigaBoost**: Most prominent
 *
 */

type Toggle = { id: string; value: BoostLevels; label: string };

// Define all possible boost levels for toggles
const TOGGLES: Record<BoostLevels, Toggle> = {
	default: {
		id: 'boostlevel-0',
		value: 'default',
		label: 'Default',
	},
	boost: {
		id: 'boostlevel-1',
		value: 'boost',
		label: 'Boost',
	},
	megaboost: {
		id: 'boostlevel-2',
		value: 'megaboost',
		label: 'Mega Boost',
	},
	gigaboost: {
		id: 'boostlevel-3',
		value: 'gigaboost',
		label: 'Giga Boost',
	},
};

/**
 * ## Presets:
 * To simplify the boosting process for editors, we provide preset groups for different collection types.
 * We override the labels in some groups so they make more sense in the context of the collection.
 *
 * ### Flexible General (4 Groups):
 * - **Group 0 (Standard)**: Default, Boost, MegaBoost
 * - **Group 1 (Big)**: Boost, MegaBoost
 * - **Group 2 (Very Big)**: MegaBoost
 * - **Group 3 (Splash)**: Default, Boost, MegaBoost, GigaBoost
 *
 * ### Flexible Special (2 Groups):
 * - **Group 0 (Standard)**: Default, Boost, MegaBoost, GigaBoost
 * - **Group 1 (Splash)**: Default, Boost, MegaBoost, GigaBoost
 *
 * */
export const CollectionToggles: Record<string, Record<number, Toggle[]>> = {
	'flexible/general': {
		0: [TOGGLES.default, TOGGLES.boost, TOGGLES.megaboost],
		1: [
			{ ...TOGGLES.boost, label: 'Default' },
			{ ...TOGGLES.megaboost, label: 'Boost' },
		],
		2: [{ ...TOGGLES.megaboost, label: 'Default' }],
		3: [TOGGLES.default, TOGGLES.boost, TOGGLES.megaboost, TOGGLES.gigaboost],
	},
	'flexible/special': {
		0: [TOGGLES.default, TOGGLES.boost, TOGGLES.megaboost, TOGGLES.gigaboost],
		1: [TOGGLES.default, TOGGLES.boost, TOGGLES.megaboost, TOGGLES.gigaboost],
	},
};

const getInputId = (cardId: string, label: string) => `${cardId}-${label}`;

/**
 * ## Function:
 * Given a `groupIndex`, `cardId`, and `collectionType`, this function returns the appropriate
 * boost level radio button options.
 */
export const renderBoostToggles = (
	groupIndex: number = 0,
	cardId: string,
	onChange: (value: string) => void,
	collectionType?: string,
) => {
	// Only render boost toggles for flexible collections
	if (
		!collectionType ||
		!['flexible/general', 'flexible/special'].includes(collectionType)
	) {
		return [];
	}

	const toggles = CollectionToggles[collectionType][groupIndex];

	return toggles.map(({ label, id, value }) => (
		<Field
			key={id}
			name={'boostLevel'}
			component={InputRadio}
			label={label}
			id={getInputId(cardId, id)}
			value={value}
			type="radio"
			onChange={() => onChange('boostLevel')}
		/>
	));
};
