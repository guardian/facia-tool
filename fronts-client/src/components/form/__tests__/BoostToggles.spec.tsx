import React from 'react';
import renderer from 'react-test-renderer';
import { renderBoostToggles } from '../BoostToggles';

// Mock redux-form Field component so that we dont need to create a full redux-form setup (store, etc)
jest.mock('redux-form', () => ({
	Field: (props: Record<string, unknown>) => (
		<div {...props} data-testid="mock-field" />
	),
}));

const mockCardId = 'test-card';

const groupToggles = [
	{
		group: 0,
		collectionType: 'flexible/general',
		expectedLabels: ['default', 'boost', 'megaboost'],
	},
	{
		group: 1,
		collectionType: 'flexible/general',
		expectedLabels: ['default', 'boost'],
	},
	{ group: 2, collectionType: 'flexible/general', expectedLabels: ['default'] },
	{
		group: 3,
		collectionType: 'flexible/general',
		expectedLabels: ['default', 'boost', 'megaboost', 'gigaboost'],
	},
	{
		group: 0,
		collectionType: 'flexible/special',
		expectedLabels: ['default', 'boost', 'megaboost'],
	},
	{
		group: 1,
		collectionType: 'flexible/special',
		expectedLabels: ['default', 'boost', 'megaboost', 'gigaboost'],
	},
];

describe('renderBoostToggles', () => {
	groupToggles.forEach(({ group, collectionType, expectedLabels }) => {
		it(`returns the correct toggles for ${collectionType} group ${group}`, () => {
			// ✅ Get the raw output instead of rendering in a form
			const toggles = renderBoostToggles(group, mockCardId, collectionType);

			// ✅ Render only the toggles, skipping redux-form & form
			const tree = renderer.create(<>{toggles}</>).toJSON();
			expect(tree).toMatchSnapshot();

			// Ensure expected labels exist
			const nodes = Array.isArray(tree) ? tree : [tree];
			const labels = nodes.map((node) => node?.props.label);
			expect(labels).toEqual(expectedLabels);
		});
	});

	it('returns an empty fragment if collectionType is invalid', () => {
		const toggles = renderBoostToggles(0, mockCardId, 'invalid/type');
		const tree = renderer.create(<>{toggles}</>).toJSON();

		expect(tree).toMatchSnapshot();
		expect(tree).toBeNull(); // Ensure it returns nothing
	});

	it('returns an empty fragment if collectionType is missing', () => {
		const toggles = renderBoostToggles(0, mockCardId);
		const tree = renderer.create(<>{toggles}</>).toJSON();

		expect(tree).toMatchSnapshot();
		expect(tree).toBeNull(); // Ensure it returns nothing
	});
});
