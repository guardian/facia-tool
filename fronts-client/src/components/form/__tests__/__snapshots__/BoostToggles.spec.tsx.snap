// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderBoostToggles returns an empty fragment if collectionType is invalid 1`] = `null`;

exports[`renderBoostToggles returns an empty fragment if collectionType is missing 1`] = `null`;

exports[`renderBoostToggles returns the correct toggles for flexible/general group 0 1`] = `
[
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-0"
    label="Default"
    name="boostLevel"
    type="radio"
    value="default"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-1"
    label="Boost"
    name="boostLevel"
    type="radio"
    value="boost"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-2"
    label="Mega Boost"
    name="boostLevel"
    type="radio"
    value="megaboost"
  />,
]
`;

exports[`renderBoostToggles returns the correct toggles for flexible/general group 1 1`] = `
[
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-1"
    label="Default"
    name="boostLevel"
    type="radio"
    value="boost"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-2"
    label="Boost"
    name="boostLevel"
    type="radio"
    value="megaboost"
  />,
]
`;

exports[`renderBoostToggles returns the correct toggles for flexible/general group 2 1`] = `
<div
  component={[Function]}
  data-testid="mock-field"
  id="test-card-boostlevel-2"
  label="Default"
  name="boostLevel"
  type="radio"
  value="megaboost"
/>
`;

exports[`renderBoostToggles returns the correct toggles for flexible/general group 3 1`] = `
[
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-0"
    label="Default"
    name="boostLevel"
    type="radio"
    value="default"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-1"
    label="Boost"
    name="boostLevel"
    type="radio"
    value="boost"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-2"
    label="Mega Boost"
    name="boostLevel"
    type="radio"
    value="megaboost"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-3"
    label="Giga Boost"
    name="boostLevel"
    type="radio"
    value="gigaboost"
  />,
]
`;

exports[`renderBoostToggles returns the correct toggles for flexible/special group 0 1`] = `
[
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-0"
    label="Default"
    name="boostLevel"
    type="radio"
    value="default"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-1"
    label="Boost"
    name="boostLevel"
    type="radio"
    value="boost"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-2"
    label="Mega Boost"
    name="boostLevel"
    type="radio"
    value="megaboost"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-3"
    label="Giga Boost"
    name="boostLevel"
    type="radio"
    value="gigaboost"
  />,
]
`;

exports[`renderBoostToggles returns the correct toggles for flexible/special group 1 1`] = `
[
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-0"
    label="Default"
    name="boostLevel"
    type="radio"
    value="default"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-1"
    label="Boost"
    name="boostLevel"
    type="radio"
    value="boost"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-2"
    label="Mega Boost"
    name="boostLevel"
    type="radio"
    value="megaboost"
  />,
  <div
    component={[Function]}
    data-testid="mock-field"
    id="test-card-boostlevel-3"
    label="Giga Boost"
    name="boostLevel"
    type="radio"
    value="gigaboost"
  />,
]
`;
